Use a minimal initrd to switch to the full initrd stored in /usr

pothos opened the pull request:
The growth of binaries over time and the inclusion of new features
    filled the available boot partition space, so that the kernel+initrd
    almost couldn't fit twice anymore as required for updates. We employed
    workarounds such as wrapper scripts for ignition, afterburn and other
    binaries so that they are loaded from /usr. However, this was still not
    enough and we would have to do the same for (network) kernel modules and
    firmware. To avoid making this ever more complex we can use a dedicated
    initrd focused on loading the full initrd from /usr and then this full
    initrd can use dracut as before and even drop all the workarounds we
    accumulated.

Introduce a busybox init script that prepares a minimal environment,
    has debug toggles and an emergency shell, and only loads the real initrd
    from /usr to switch over to it. Because mdev is not a proper udev
    replacement, some additional scripting is needed. Busybox's modprobe
    can't work with dependencies well and we need the real kmod for that
    (which is also good to guarantee have the same modprobe options set).
    Also, some other busybox commands are often lacking things such as
    loading a kernel module automatically and this has to be done
    explicitly. We still set up dm-verity for /usr so that we have the same
    security properties (The code comes from the bootengine systemd
    generators we have and also covers the PXE boot with a squashfs /usr
    passed from an additional cpio). The real initrd then reuses the mount
    point for /usr, and loads any kernel modules and firmware that wasn't
    loaded already.
    We also have to make the dependencies for parse-ip-for-networkd.service
    a bit more explicit because the removal of the /sysusr mount in the full
    initrd exposed a race condition.

## How to use

With https://github.com/flatcar/scripts/pull/3241

## Testing done

See above

pothos committed:
Use a minimal initrd to switch to the full initrd stored in /usr

The growth of binaries over time and the inclusion of new features
filled the available boot partition space, so that the kernel+initrd
almost couldn't fit twice anymore as required for updates. We employed
workarounds such as wrapper scripts for ignition, afterburn and other
binaries so that they are loaded from /usr. However, this was still not
enough and we would have to do the same for (network) kernel modules and
firmware. To avoid making this ever more complex we can use a dedicated
initrd focused on loading the full initrd from /usr and then this full
initrd can use dracut as before and even drop all the workarounds we
accumulated.

Introduce a busybox init script that prepares a minimal environment,
has debug toggles and an emergency shell, and only loads the real initrd
from /usr to switch over to it. Because mdev is not a proper udev
replacement, some additional scripting is needed. Busybox's modprobe
can't work with dependencies well and we need the real kmod for that
(which is also good to guarantee have the same modprobe options set).
Also, some other busybox commands are often lacking things such as
loading a kernel module automatically and this has to be done
explicitly. We still set up dm-verity for /usr so that we have the same
security properties (The code comes from the bootengine systemd
generators we have and also covers the PXE boot with a squashfs /usr
passed from an additional cpio). The real initrd then reuses the mount
point for /usr, and loads any kernel modules and firmware that wasn't
loaded already.
We also have to make the dependencies for parse-ip-for-networkd.service
a bit more explicit because the removal of the /sysusr mount in the full
initrd exposed a race condition.

Signed-off-by: Kai Lueke <kailuke@microsoft.com>

krnowak commented on minimal-init:
```diff
@@ -0,0 +1,189 @@
+}
+
+emergency() {
+  echo "ERROR: The early initrd has failed. To active debug shell breakpoints, boot with rd.earlyshell in the kernel cmdline, and to active tracing, boot with rd.earlytrace" >&2
```
Did you mean to use `activate` here?

```suggestion
  echo "ERROR: The early initrd has failed. To activate debug shell breakpoints, boot with rd.earlyshell in the kernel cmdline, and to activate tracing, boot with rd.earlytrace" >&2
```

pothos replied:
Yes, typo, thanks

krnowak commented on minimal-init:
```diff
@@ -0,0 +1,189 @@
+    # If problems arise, first make sure that required modules and their deps are actually in the initrd,
+    # but if that's not enough we might even have to trigger the find /sys ... xargs coldplugging
+    # here again every now and then? (Last resort would be to run proper udev, possibly without systemd.)
+    if [ $((now - starttime)) -gt 300 ]; then
```
Would it make sense to be able to override the timeout through the cmdline parameters? Dunno, maybe there are platforms, where you need to wait for the disk to show up.

pothos replied:
Yes, we could add a parameter, I thought 5 minutes are very long already but who knows, and it could be handy when testing to set it to a smaller timout

pothos replied:
Added `rd.earlytimeout`

krnowak commented on minimal-init:
```diff
@@ -0,0 +1,189 @@
+# Only proceed if the source is a path and we have sufficient parameters.
+if echo "${verityusr}" | grep -q "^/" && [ "${usrhash}" != "" ]; then
+  # Hardcoded expected value from the image GPT layout
+  veritysetup --panic-on-corruption --hash-offset=1065345024 open "${verityusr}" usr "${verityusr}" "${usrhash}"
```
Weren't offsets different on amd64 and arm64? Or did I just confuse things?

pothos replied:
For the place where the hash is located in the kernel binary, yes, but the verity data offset on disk is the same.

krnowak commented on minimal-init:
```diff
@@ -0,0 +1,189 @@
+echo "Mounting /usr from ${usr}" >&2
+# mount -t auto only works if btrfs is already loaded
+modprobe btrfs
+mount -t "${usrfstype}" -o "${usrflags}" "${usr}" /sysusr/usr
```
Hm, will `mount` create the `/sysusr/usr` directories for us or it is created elsewhere?

pothos replied:
It's set up in the initrd build step, same for the `realinit` folder. We could also add `mkdir -p` here again?

krnowak replied:
Doing `mkdir -p /sysusr/usr` certainly wouldn't hurt.

pothos replied:
Sure, added it, also for `/realinit`

krnowak reviewed (approved):
Looks good, but I think Chewi will want to have a look too.

chewi commented on minimal-init:
```diff
@@ -0,0 +1,194 @@
+busybox mount -n -t proc proc /proc
+busybox mount -n -t devtmpfs devtmpfs /dev
+busybox mount -n -t sysfs sysfs /sys
+busybox --install -s
```
Perhaps we still need this for subprocesses, but the busybox shell does not need symlinks to invoke built-in tools without prepending them with `busybox`. It will prefer a tool from the PATH if one exists, otherwise it will use the built-in version.

pothos replied:
I think there's a config option for busybox that selects how built-in execution is done. Anyway, the --install -s line sets up the symlinks and afterwards everything is prepared to also be reachable through the PATH.

chewi commented on minimal-init:
```diff
@@ -0,0 +1,194 @@
+mdev -s
+# Coldplugging but with using /sbin/modprobe (which is kmod) instead of busybox's modprobe
+# because busybox doesn't properly support the globs in modules.alias
+find /sys/ -name modalias -print0 | xargs -0 sort -u | tr '\n' '\0' | xargs -0 /sbin/modprobe -abq || true
```
What gets passed to modprobe exactly? When I try this on my systems, I get entries like this, which `modprobe` doesn't understand.

```
of:NvopT(null)Crockchip,rk3566-vop
of:NwatchdogT(null)Crockchip,rk3568-wdtCsnps,dw-wdt
pci:v00001D87d00003566sv00000000sd00000000bc06sc04i00
pci:v00001E0Fd00000009sv00001E0Fsd00000001bc01sc08i02
platform:alarmtimer
```

pothos replied:
The modules.alias file maps these to module names, if a module exists for it.

chewi commented on minimal-init:
```diff
@@ -0,0 +1,194 @@
+    fi
+    # No "sleep 0.1", so this is rather busy polling
+    if [ "${ueventline}" != "" ]; then
+      drive="$({ grep -s -l -m 1 -r "${ueventline}" /sys/class/block/*/uevent || true; } | cut -d / -f 5)"
+    else
+      drive="$(blkid | { grep -m 1 "${blkidmatch}" || true ; } | cut -d : -f 1 | cut -d / -f 3-)"
```
I don't think you need the `|| true` because there is no `set -o pipefail`.

```suggestion
      drive="$(grep -s -l -m 1 -r "${ueventline}" /sys/class/block/*/uevent | cut -d / -f 5)"
    else
      drive="$(blkid | grep -m 1 "${blkidmatch}" | cut -d : -f 1 | cut -d / -f 3-)"
```

pothos replied:
Yeah, but one one adds it (or copies this line to another script), this breaks, so I think we better have this.

chewi replied:
busybox doesn't support it, and I would say that if anyone copies this code for use outside busybox, then they're probably doing something bad. :stuck_out_tongue: Your call!

pothos replied:
A recent enough busybox supports it now. I haven't set `-eo pipefail` up here though because of the emergency shell trap that we have.

chewi commented on minimal-init:
```diff
@@ -0,0 +1,194 @@
+# Only proceed if the source is a path and we have sufficient parameters.
+if echo "${verityusr}" | grep -q "^/" && [ "${usrhash}" != "" ]; then
+  # Hardcoded expected value from the image GPT layout
+  veritysetup --panic-on-corruption --hash-offset=1065345024 open "${verityusr}" usr "${verityusr}" "${usrhash}"
```
Bottlerocket was able to do this without `veritysetup` at all by passing this on the kernel command line! You could do the same with `modprobe`. Or if you still need `dmsetup`, you could use that.

```
dm-mod.create="root,,,ro,0 1884160 verity 1 PARTUUID=$boot_uuid/PARTNROFF=1 PARTUUID=$boot_uuid/PARTNROFF=2 4096 4096 235520 1 sha256 8b461e3a50dfdf97d78d8d52e90772bb44c9d3b376d7f662adaea98b8202e57e 808fcbd183fa45e9fb787fcc4732e39eb70db61e26d4b21ba6896717ebe9d4f9 2 restart_on_corruption ignore_zero_blocks"
```

This probably won't be enough if we decide to sign the verity hash, but I think that's off the table now.

chewi replied:
On reflection, I guess `veritysetup` isn't a heavy requirement.

pothos replied:
With the libs it pulls in it is considerable but since the overall size wins are big enough I accepted this cost. Interesting that these dm-mod parameters are also possible!

chewi commented on minimal-init:
```diff
@@ -0,0 +1,194 @@
+if echo "${usr}" | grep -v -q "^/"; then
+  echo "No mountable /usr partition given (usr='${usr}')" >&2
+  false # Throw error
+fi
```
```suggestion
case "${usr}" in
  /*) : ;;
  *) echo "No mountable /usr partition given (usr='${usr}')" >&2
    false # Throw error
    ;;
esac
```

pothos replied:
Thanks

chewi reviewed (changes requested):
Well, it's far more comprehensive than what I was shooting for. I guess I shouldn't assume that it will always just work!

Was there any particular reason why you went with squashfs over erofs in the end, other than mkfs.erofs not being in the SDK? It probably should be.

Even without these changes, the verity-generator module and much of the usr-generator module is redundant anyway following changing in systemd. I have already dropped that code in my sd-veritysetup branch, which was originally going to be part of my boot rework, but I was also planning to apply it regardless. Generally speaking, I don't like leaving dead code around. That's what git is for. :wink:

I recall someone saying something about our existing boot process being susceptible to duplicate disks, where it might pick USR-A from a disk other than the one you actually booted from. Maybe we could address that here (later?) although I'm not sure how you would tell which was the boot disk.

pothos commented:
> Was there any particular reason why you went with squashfs over erofs in the end, other than mkfs.erofs not being in the SDK? It probably should be.

So far we use squashfs for everything and I kept using it here for consistency. If we do it, I would prefer a follow-up PR.

> Even without these changes, the verity-generator module and much of the usr-generator module is redundant anyway following changing in systemd. I have already dropped that code in my sd-veritysetup branch, which was originally going to be part of my boot rework, but I was also planning to apply it regardless. Generally speaking, I don't like leaving dead code around. That's what git is for. 😉

Yes, good point. The generators don't handle the missing `mount.` prefix, or - but that could also be gained by preprocessing? I kept the files around because I thought we might go back to this - either now when things don't work on real hardware or later if we have enough space again - but I'm also ok with removing entirely and relying on git.

pothos committed:
Remove unused generator code

The minimal initrd now does the verity setup and /sysusr mounting and we
can delete parts of the generator code.
Delete the unused code but still keep support for "usr=" instead of
"mount.usr=" cmdline entries. We could decide to remove this later to
get rid of a few more things in the generators. Still, /sysroot(/usr)
setup will be needed. For "mount.usr=" the upstream systemd generator
handles this but there's still a bit of code to customize the mount args
for it. The verity generator gets deleted fully and if we later want to
use it again, we can use the upstream one with some cmdline
preprocessing.

Signed-off-by: Kai Lueke <kailuke@microsoft.com>

chewi commented on minimal-init:
```diff
@@ -0,0 +1,196 @@
+mkdir -p /realinit/sysusr/usr
+mount -o move /sysusr/usr /realinit/sysusr/usr
+if [ "${usr}" = /usr.squashfs ]; then
+  mkdir -p /oem
+  mkdir -p /realinit/oem
+  mount -o bind /oem /realinit/oem
```
@pothos, do you remember why you used `/oem` rather than `/usr/share/oem` here? Our stock initrd doesn't have the former, and the latter would have avoided flatcar/Flatcar#2023. With further changes, I could reduce our dependence on the legacy `/usr/share/oem` path, but I feel like I'm missing something here.

pothos replied:
I think this topic came also up when we moved from `/usr/share/oem` to `/oem` because this also broke there?

I think we should do a check here for the path contents of both `/oem` and `/usr/share/oem` and use the first one that has something.

pothos replied:
If the bind mount is the problem we could also try a `mv` to be sure that we don't use more RAM than needed?

Needs testing with the pxe qemu script again.

chewi replied:
I think I'll just move `/usr/share/oem/*` to `/oem` first or something like that.  The other references to `/usr/share/oem` in bootengine can be changed to `/oem`. I don't think the bind mount is an issue, but I'll check.
